### DISCUSSION TITLE
Debug not working

### DISCUSSION THREAD

**Discussion by PerennialNovice** [stm32duino/Arduino_Core_STM32#2603] [Q&A] [answered]
when trying to debug on a fresh (Windows) install of Arduino IDE 2.3.4 and Srduino_Core_STM32 2.9.0 on Nucleo L4R5ZI I get the following error message on a first debug attempt

```
xPack Open On-Chip Debugger 0.12.0+dev-01685-gb9224c0c0-dirty (2024-08-02-19:51)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
DEPRECATED! use 'gdb port', not 'gdb_port'
DEPRECATED! use 'tcl port' not 'tcl_port'
CDRTOSConfigure
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : Listening on port 50001 for tcl connections
Info : Listening on port 50002 for telnet connections
Info : clock speed 500 kHz
Info : STLINK V2J45M31 (API v2) VID:PID 0483:374B
Info : Target voltage: 3.239216
Error: init mode failed (unable to connect to the target)


[2024-12-12T11:04:05.909Z] SERVER CONSOLE DEBUG: onBackendConnect: gdb-server session closed
GDB server session ended. This terminal will be reused, waiting for next session to start...
```

I then realized I hadn't set the debug options and also reset the board

on every subsequent debug attempt, I then get:

```
xPack Open On-Chip Debugger 0.12.0+dev-01685-gb9224c0c0-dirty (2024-08-02-19:51)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
DEPRECATED! use 'gdb port', not 'gdb_port'
DEPRECATED! use 'tcl port' not 'tcl_port'
CDRTOSConfigure
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : Listening on port 50001 for tcl connections
Info : Listening on port 50002 for telnet connections
Info : clock speed 500 kHz
Error: libusb_open() failed with LIBUSB_ERROR_NOT_FOUND
Error: open failed


[2024-12-12T11:19:37.914Z] SERVER CONSOLE DEBUG: onBackendConnect: gdb-server session closed
GDB server session ended. This terminal will be reused, waiting for next session to start...
```

**Comment by PerennialNovice** [comment]
after a rebuild of the Blink example, I now have a combination of both errors

```
xPack Open On-Chip Debugger 0.12.0+dev-01685-gb9224c0c0-dirty (2024-08-02-19:51)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
DEPRECATED! use 'gdb port', not 'gdb_port'
DEPRECATED! use 'tcl port' not 'tcl_port'
CDRTOSConfigure
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : Listening on port 50001 for tcl connections
Info : Listening on port 50002 for telnet connections
Info : clock speed 500 kHz
Error: libusb_open() failed with LIBUSB_ERROR_NOT_FOUND
Info : STLINK V2J45M31 (API v2) VID:PID 0483:374B
Info : Target voltage: 3.239765
Error: init mode failed (unable to connect to the target)


[2024-12-12T11:35:42.381Z] SERVER CONSOLE DEBUG: onBackendConnect: gdb-server session closed
GDB server session ended. This terminal will be reused, waiting for next session to start...
```

**Reply by fpistm** [reply]
Install libusb for your OS

**Reply by PerennialNovice** [reply]
will do

... but how can it be that it is needed after debug worked well for "ages" ?
Is this a change in 2.9.0 ?

**Reply by fpistm** [reply]
Only xpack openocd. You should open an issue on the dedicated xpack repo. To warn the maintainer there is a change not expected vs previous release.

**Reply by fpistm** [reply]
I didn't saw this issue as I've the libusb installed on Windows and Linux.

**Comment by PerennialNovice** [comment]
So I guess this was some kind of hiccup specific to my machine/setup.

But what is strange (at least to me), is that I can program the MCU with the very same setup that fails when trying to debug...

**Comment by PerennialNovice** [comment]
I just tried using openOCD STLink for programming - which failed as well
so at least it is consistent to openOCD not being able to talk to my STM32L412
I have no problem debugging STM32L4R5

**Reply by fpistm** [reply]
Very strange it failed only for STM32L412 and not for other.
Seems related to OpenOCD + selected target in that case. But can't help on this third party.

**Reply by fpistm** [reply]
Is it a Nucleo L412? If yes try to update the STLink firmware with the STM32CubeProgrammer.

**Comment by PerennialNovice** [comment]
I used STLINK-V3MINIE to connect to both custom boards. It has V3J15M7 firmware, which according to STM32CubeProgrammer 2.18.0 is the latest.

Luckily I have Nucleo-L4R5 and Nucleo-L412 at hands. Both have V2J45M31 firmware, which should be the latest.

For Nucleo-L4R5, I first got a `Error: init mode failed (unable to connect to the target)` when trying to debug. After programming it normally, I was able to debug it successfully.

But Nucleo-L412 always threw me an `Error: init mode failed (unable to connect to the target)`. Also when trying to program it via openOCD STLink (SWD). Programming via STM32CubeProgrammer (SWD) works fine.

It definitely seems to be linked to the L412 somehow.

Little side-question: to me it seems that starting debug does neither recompile, nor download the binary to the board. Is this observation correct?

**Reply by fpistm** [reply]
Ah ok, with L412 it is not the libusb issue. 
This is the same config for the Nucleo-L4R5 and  Nucleo-L412. Maybe the hardware reset if no functional with this board.
You can try to comment this line in the stm32l4x.cfg in the `C:\Users\<username>\AppData\Local\Arduino15\packages\STMicroelectronics\tools\xpack-openocd\0.12.0-4\openocd\scripts\target\stm32l4x.cfg` :
`reset_config srst_nogate`

> Little side-question: to me it seems that starting debug does neither recompile, nor download the binary to the board. Is this observation correct?

Right, you have to upload it before start debugging. Debug does not manage upload. So if you made a change on your code, build, upload then debug.

**Comment by PerennialNovice** [comment]
modifiying the config allows me to connect the debugger, but it does not seem to work.
At first I could not halt the target, I then did a restart (using the green debugging button)
Then I can halt the target, but it does not jump to the right place in the code and it looks like the debugger cannot find the correct functions

![image](https://github.com/user-attachments/assets/0b9b50ca-bf3f-4a86-8daf-5d3e038a9a68)

**Reply by fpistm** [reply]
You have to enable the "Optimize for debugging" in the scketch menu else you don't have the symbol. And note that you have to always use the green button... Arduino IDe restriction

**Comment by PerennialNovice** [comment]
which green button?
![image](https://github.com/user-attachments/assets/f2e37dc3-6387-4174-8c89-0714617d8675)

this is how the board is configured:
![image](https://github.com/user-attachments/assets/e1ee2666-204b-491e-921b-98f6dce7771c)

optimize for debug is ticked:
![image](https://github.com/user-attachments/assets/ee32dadd-1ce0-4648-9799-49c4809df331)

**Reply by fpistm** [reply]
The round one close to upload

**Reply by PerennialNovice** [reply]
I did all as you said, but still cannot debug the Nucleo-L412

**Reply by fpistm** [reply]
Will test next week if I found this board but sounds very strange as I've already debug this one and the config is correct.

**Reply by PerennialNovice** [reply]
Nice, thanks!
I will try to get hands on another board, or order a new one just to rule out it is a hardware defect of some kind

**Reply by fpistm** [reply]
Hi @PerennialNovice 
I do not have a Nucleo L412KB but tested with Nucleo-L432KC which have the same reference board (MB1180) and it works as expected.
Using same generic openocd config.
Also tested the Nucleo-L4R5 and works as expected without remove the reset feature.
Did you modified the board? Remove/add SB?
Unfortunately, could not help more as it seems issue is only on your side.

Maybe a driver issue of the STLink?

**Comment by PerennialNovice** [comment]
I am waiting for a brandnew Nucleo L412 to arrive and I might have a L432 lying around at home
I will be able to say more tomorrow

**Comment by PerennialNovice** [comment]
the brandnew Nucleo-L412KB works as it should
another L412 I found works as well
Nucleo L432KC also works as it should

So I have one Nucleo-L412 that I am not able to debug and 2 custom board designs that I cannot debug
the undebuggable Nucleo shows signs of soldering
Looks like I'm in for bug hunting then ;)

Thanks for your support

**Comment by PerennialNovice** [comment, marked as answer]
@fpistm well, for the Nucleo-L412KB you were absolutely on the right track!
SB12 was open, which deconnects BOOT0 

soldering it back restored debug capabilities :)

Thanks again!

Now, I "only" have to find out what is wrong with my custom design. I had a sharp look at the Nucleo schematic before creating it, not to say that I copied a lot from there ;P

**Reply by fpistm** [reply]
Fine.😉
